### ISSUE TITLE
SDPA Attention Backend

### ISSUE THREAD

**Issue by keyu-wang-2002** [zjunlp/LightThinker#8]
Hi, nice work, and thanks for the clean codebase.

I noticed that you implemented both FlashAttention and SDPA Attention in the model. SDPA Attention has three modes: Flash, Efficient, and Math. For the base/vanilla model, standard causal (lower-triangular) attention can benefit from FlashAttention and run through the SDPA Flash path. However, LightThinker uses sparse attention, which does not support FlashAttention or SDPA Flash, so it can only run through SDPA Efficient, which is noticeably slower than Flash. I also confirmed this by printing backend info during inference. (Maybe FlexAttention suits LightThinker better?..)

In my reproduction, my trained Vanilla is faster than LightThinker:

| Model/Task | gsm8k (acc/time) | mmlu (acc/time) | gpqa (acc/time) | bbh (acc/time) | 
|---------------------|------------------|-----------------|-----------------|----------------| 
| vanilla r1-distill-qwen | 89.91 / 12.65 | 62.51 / 16.12 | 35.86 / 9.98 | 45.05 / 8.80 | 
| qwen_tho1_comp9 | 85.29 / 14.80 | 61.05 / 16.02 | 25.76 / 10.44 | 43.43 / 9.70 | 
| qwen_tok6_comp2 | 83.01 / 14.53 | 56.77 / 16.91 | 27.78 / 9.69 | 39.19 / 10.01 | | qwen_tok4_comp2 | 84.38 / 14.71 | 57.74 / 14.96 | 26.26 / 8.90 | 40.20 / 9.52 |
| qwen_tok4_comp2 | 84.38 / 14.71 | 57.74 / 14.96 | 26.26 / 8.90 | 40.20 / 9.52 |

All models are from the Qwen2.5-7B series. Here, qwen_tho1_comp9 corresponds to Ours (tho.), and qwen_tok6_comp2 corresponds to Ours (token). I ran these experiments on 4× A100 GPUs, and the configs match Appendix C. Please ignore the abnormal BBH accuracy — it’s due to output-format / regex issues and is easy to fix (and also other accuracy mismatch).

What confuses me is that the runtime of my vanilla model is much shorter than the results reported in the paper. One possible reason is that my vanilla inference uses SDPA Flash, while the paper’s vanilla runs may have disabled SDPA Flash / used SDPA Efficient instead (thought the head_dim=128 for the base model, means Flash may not perform superior than Efficient, maybe other backend used in the original paper?). Have you noticed this issue? I’d like to understand what I might be missing.

Thanks again for your work.

**Comment by MikeDean2367** [comment]
Hi, thanks for the interest in our work and the clean reproduction!

**1. Regarding Efficiency and Inference Speed**
You are spot on, the runtime difference you observed is indeed because our reported Vanilla baseline did not utilize FlashAttention (SDPA Flash).

Due to specific environment configurations during our experiments, we ran **both** the Vanilla baseline and LightThinker in **eager attention mode** (disabling Flash/SDPA optimization) to ensure a controlled comparison. This explains why your Vanilla reproduction (with Flash enabled) is significantly faster than the results in our paper.

The specific implementation path can be seen here:
https://github.com/zjunlp/LightThinker/blob/47d5c38fad315da33baa52b2763f571048f11d3e/LightThinker/model_qwen.py#L595-L599

You can also verify the active attention backend by adding a print statement or a try-except block in the `__init__` methods of `Qwen2FlashAttention2` and `Qwen2SdpaAttention` to see which one initializes successfully.
```python
class Qwen2FlashAttention2(Qwen2Attention):
    def __init__(self, *args, **kwargs):
        try:
            super().__init__(*args, **kwargs)
            self._flash_attn_uses_top_left_mask = not is_flash_attn_greater_or_equal_2_10()
        except Exception as e:
            print(f"use flash attention failed: {e}")

class Qwen2SdpaAttention(Qwen2Attention):
    def __init__(self, *args, **kwargs):
        try:
            super().__init__(*args, **kwargs)
        except Exception as e:
            print(f"use sdpa attention failed: {e}")
```

*Note on FlexAttention:* That's a great suggestion. Since LightThinker doesn't explicitly rely on the full attention score matrix and attention mask during inference, integrating something like FlashAttention and  FlexAttention could indeed bridge the gap. It is noted that FlexAttention is [available](https://pytorch.org/blog/flexattention/) in torch==2.5.0. However, we use torch==2.1.0 for experiments.

**2. Regarding Accuracy:**
For the performance discrepancies (especially the drop you noticed), I recommend evaluating checkpoints from earlier epochs (e.g., Epoch 4 or 5). We observed that the model tends to suffer from overfitting in later epochs, which hurts generalization on these benchmarks.

Let me know if you have further questions!

**Comment by keyu-wang-2002** [comment]
Thanks! No more questions about this attention backend issue. BTW happy new year :)
